src/libteam/0005-libteam-Add-warm_reboot-mode.patch

```diff
@@ -0,0 +1,354 @@
++
+ void teamd_run_loop_restart(struct teamd_context *ctx)
+ {
+ 	teamd_run_loop_sent_ctrl_byte(ctx, 'r');
+@@ -694,6 +716,10 @@ static int callback_daemon_signal(struct teamd_context *ctx, int events,
+ 		teamd_log_warn("Got SIGINT, SIGQUIT or SIGTERM.");
+ 		teamd_run_loop_quit(ctx, 0);
+ 		break;
++	case SIGUSR1:
++		teamd_log_warn("Got SIGUSR1.");
++		teamd_run_loop_quit_w_boot(ctx, 0);
++		break;
+ 	}
+ 	return 0;
+ }
+@@ -1507,7 +1533,7 @@ static int teamd_start(struct teamd_context *ctx, enum teamd_exit_code *p_ret)
+ 		return -errno;
+ 	}
+ 
+-	if (daemon_signal_init(SIGINT, SIGTERM, SIGQUIT, SIGHUP, 0) < 0) {
++	if (daemon_signal_init(SIGINT, SIGTERM, SIGQUIT, SIGHUP, SIGUSR1, 0) < 0) {
+ 		teamd_log_err("Could not register signal handlers.");
+ 		daemon_retval_send(errno);
+ 		err = -errno;
+diff --git a/teamd/teamd.h b/teamd/teamd.h
+index ef0fb1c..503d9db 100644
+--- a/teamd/teamd.h
++++ b/teamd/teamd.h
+@@ -125,6 +125,9 @@ struct teamd_context {
+ 	char *				hwaddr;
+ 	uint32_t			hwaddr_len;
+ 	bool				hwaddr_explicit;
++	bool				warm_reboot;
++	bool				keep_ports;
++	char *				lacp_directory;
+ 	struct {
+ 		struct list_item		callback_list;
+ 		int				ctrl_pipe_r;
+@@ -191,12 +194,15 @@ struct teamd_event_watch_ops {
+ 				   struct teamd_port *tdport, void *priv);
+ 	void (*refresh)(struct teamd_context *ctx,
+                    struct teamd_port *tdport, void *priv);
++	void (*port_flush_data)(struct teamd_context *ctx,
++                   struct teamd_port *tdport, void *priv);
+ 	int (*option_changed)(struct teamd_context *ctx,
+ 			      struct team_option *option, void *priv);
+ 	char *option_changed_match_name;
+ };
+ 
+ void teamd_refresh_ports(struct teamd_context *ctx);
++void teamd_ports_flush_data(struct teamd_context *ctx);
+ int teamd_event_port_added(struct teamd_context *ctx,
+ 			   struct teamd_port *tdport);
+ void teamd_event_port_removed(struct teamd_context *ctx,
+diff --git a/teamd/teamd_events.c b/teamd/teamd_events.c
+index 5c2ef56..50e5a08 100644
+--- a/teamd/teamd_events.c
++++ b/teamd/teamd_events.c
+@@ -47,6 +47,19 @@ void teamd_refresh_ports(struct teamd_context *ctx)
+ 	}
+ }
+ 
++void teamd_ports_flush_data(struct teamd_context *ctx)
++{
++	struct teamd_port *tdport;
++	struct event_watch_item *watch;
++
++	teamd_for_each_tdport(tdport, ctx) {
++		list_for_each_node_entry(watch, &ctx->event_watch_list, list) {
++			if (!watch->ops->port_flush_data) continue;
++			watch->ops->port_flush_data(ctx, tdport, watch->priv);
++		}
++	}
++}
++
+ int teamd_event_port_added(struct teamd_context *ctx,
+ 			   struct teamd_port *tdport)
+ {
+diff --git a/teamd/teamd_runner_lacp.c b/teamd/teamd_runner_lacp.c
+index 81324de..45d667d 100644
+--- a/teamd/teamd_runner_lacp.c
++++ b/teamd/teamd_runner_lacp.c
+@@ -174,6 +174,9 @@ struct lacp_port {
+ 	struct lacp_port *agg_lead; /* leading port of aggregator.
+ 				     * NULL in case this port is not selected */
+ 	enum lacp_port_state state;
++	bool lacpdu_received;
```

Comment:
Didn't see it being used.

Reply:
Yes. You're absolutely right. I've found this issue too, but didn't have a chance to fix it. I'm going to fix it on the next iteration.

Reply:
Fixed